projects/rocprofiler-systems/source/lib/core/trace_cache/buffer_storage.cpp

```diff
@@ -0,0 +1,171 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "buffer_storage.hpp"
+#include "debug.hpp"
+#include "library/runtime.hpp"
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <unistd.h>
+
+using namespace std::chrono_literals;
+
+namespace rocprofsys
+{
+namespace trace_cache
+{
+
+namespace
+{
+constexpr auto CACHE_FILE_FLUSH_TIMEOUT = 10ms;
+}  // namespace
+
+buffer_storage::buffer_storage(pid_t _pid)
+{
+    ROCPROFSYS_SCOPED_SAMPLING_ON_CHILD_THREADS(false);
+    m_flushing_thread = std::make_unique<std::thread>([this, _pid]() {
+        std::ofstream _ofs(filename, std::ios::binary | std::ios::out);
+
+        if(!_ofs)
+        {
+            std::stringstream _ss;
+            _ss << "Error opening file for writing: " << filename;
+            throw std::runtime_error(_ss.str());
+        }
+
+        auto execute_flush = [&](std::ofstream& ofs, bool force = false) {
+            size_t _head, _tail;
+            {
+                std::lock_guard guard{ m_mutex };
+                _head = m_head;
+                _tail = m_tail;
+
+                if(_head == _tail)
+                {
+                    return;
+                }
+
+                auto used_space =
+                    m_head > m_tail ? (m_head - m_tail) : (buffer_size - m_tail + m_head);
+                if(!force && used_space < flush_threshold)
+                {
+                    return;
+                }
+                m_tail = m_head;
+            }
+
+            if(_head > _tail)
+            {
+                ofs.write(reinterpret_cast<const char*>(m_buffer->data() + _tail),
+                          _head - _tail);
+            }
+            else
+            {
+                ofs.write(reinterpret_cast<const char*>(m_buffer->data() + _tail),
+                          buffer_size - _tail);
+                ofs.write(reinterpret_cast<const char*>(m_buffer->data()), _head);
+            }
+        };
+
+        ROCPROFSYS_DEBUG("Starting buffered storage flushing thread for pid %d",
+                         static_cast<int>(_pid));
+        m_created_process = _pid;
+        std::mutex _shutdown_condition_mutex;
+        while(m_running)
+        {
+            execute_flush(_ofs);
+            std::unique_lock _lock{ _shutdown_condition_mutex };
+            m_shutdown_condition.wait_for(
+                _lock, std::chrono::milliseconds(CACHE_FILE_FLUSH_TIMEOUT),
+                [&]() { return !m_running; });
+        }
```

Comment:
If you were using PTL, you could just submit a task to flush every X milliseconds. Or (even better), use a watermark and flush whenever a certain amount of the buffer gets filled.